drivers/dht/dht.c

```diff
@@ -67,79 +95,118 @@ int dht_init(dht_t *dev, const dht_params_t *params)
-           ((dev->type == DHT11) || (dev->type == DHT22) || (dev->type == DHT21)));
+           ((params->type == DHT11) || (params->type == DHT22) || (params->type == DHT21)));
 
-    *dev = *params;
+    memset(dev, 0, sizeof(dht_t));
+    dev->params = *params;
 
-    gpio_init(dev->pin, GPIO_OUT);
-    gpio_set(dev->pin);
+    _reset(dev);
 
     xtimer_usleep(2000 * US_PER_MS);
 
     DEBUG("dht_init: success\n");
     return DHT_OK;
 }
 
-int dht_read(const dht_t *dev, int16_t *temp, int16_t *hum)
+int dht_read(dht_t *dev, int16_t *temp, int16_t *hum)
 {
-    uint8_t csum, sum;
+    uint16_t csum, sum;
     uint16_t raw_hum, raw_temp;
 
-    assert(dev && temp && hum);
-
-    /* send init signal to device */
-    gpio_clear(dev->pin);
-    xtimer_usleep(20 * US_PER_MS);
-    gpio_set(dev->pin);
-    xtimer_usleep(40);
-
-    /* sync on device */
-    gpio_init(dev->pin, dev->in_mode);
-    while (!gpio_read(dev->pin)) {}
-    while (gpio_read(dev->pin)) {}
-
-    /*
-     * data is read in sequentially, highest bit first:
-     *  40 .. 24  23   ..   8  7  ..  0
-     * [humidity][temperature][checksum]
-     */
-
-    /* read the humidity, temperature, and checksum bits */
-    raw_hum = read(dev->pin, 16);
-    raw_temp = read(dev->pin, 16);
-    csum = (uint8_t)read(dev->pin, 8);
-
-    /* set pin high again - so we can trigger the next reading by pulling it low
-     * again */
-    gpio_init(dev->pin, GPIO_OUT);
-    gpio_set(dev->pin);
-
-    /* validate the checksum */
-    sum = (raw_temp >> 8) + (raw_temp & 0xff) + (raw_hum >> 8) + (raw_hum & 0xff);
-    if ((sum != csum) || (csum == 0)) {
-        DEBUG("error: checksum invalid\n");
-        return DHT_NOCSUM;
-    }
-
-    /* parse the RAW values */
-    DEBUG("RAW values: temp: %7i hum: %7i\n", (int)raw_temp, (int)raw_hum);
-    switch (dev->type) {
-        case DHT11:
-            *temp = (int16_t)((raw_temp >> 8) * 10);
-            *hum = (int16_t)((raw_hum >> 8) * 10);
-            break;
-        case DHT22:
-            *hum = (int16_t)raw_hum;
-            /* if the high-bit is set, the value is negative */
-            if (raw_temp & 0x8000) {
-                *temp = (int16_t)((raw_temp & ~0x8000) * -1);
+    assert(dev);
+
+    uint32_t now_us = xtimer_now_usec();
+    if ((now_us - dev->last_read_us) > DATA_HOLD_TIME) {
+        /* send init signal to device */
+        gpio_clear(dev->params.pin);
+        xtimer_usleep(20 * US_PER_MS);
+        gpio_set(dev->params.pin);
+        xtimer_usleep(40);
+
+        /* sync on device */
+        gpio_init(dev->params.pin, dev->params.in_mode);
+        unsigned counter = 0;
+        while (!gpio_read(dev->params.pin)) {
+            if (counter++ > TIMEOUT) {
+                _reset(dev);
+                return DHT_TIMEOUT;
             }
-            else {
-                *temp = (int16_t)raw_temp;
+            xtimer_usleep(1);
+        }
+
+        counter = 0;
+        while (gpio_read(dev->params.pin)) {
```

Comment:
the more I think about, maybe put this loop into its own (private) function something like:

```C
static int _sync(gpio_t pin, bool set)
{
    unsigned counter = TIMEOUT;
    while (((gpio_read(pin) > 0) == set) && counter--) {
        xtimer_usleep(1);
    }
    if (!counter) {
        return -1;
    }
    return 0;
}
```

not sure about the first loop condition, but you get the idea I guess

Reply:
I mean its used 4 times or so

Reply:
Going through your comments I had exactly the same thought :-D